File: src/34.在排序数组中查找元素的第一个和最后一个位置.cpp
/*
 * @lc app=leetcode.cn id=34 lang=cpp
 *
 * [34] 在排序数组中查找元素的第一个和最后一个位置
 */

// @lc code=start
#include <iostream>
#include <vector>
using namespace std;
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target)
    {
        int leftBorder = getLeftBorder(nums, target);
        int rightBorder = getRightBorder(nums, target);
        // 情况一
        if (leftBorder == -2 || rightBorder == -2)
            return { -1, -1 };
        // 情况三
        if (rightBorder - leftBorder >= 0)
            return { leftBorder, rightBorder };
        // 情况二
        return { -1, -1 };
    }

private:
    int getRightBorder(vector<int>& nums, int target)
    {
        int left = 0;
        int right = nums.size() - 1;
        int rightBorder = -2; // 记录一下rightBorder没有被赋值的情况
        while (left <= right) {
            int middle = left + ((right - left) / 2);
            if (nums[middle] > target) {
                right = middle - 1;
            } else if (nums[middle] < target) {
                left = middle + 1;
            } else { //当middle对应着target的时候，由于是寻找右边界，所以给left+1，然后将left赋值给rightBorder
                left = middle + 1;
                rightBorder = left - 1;
            }
        }
        return rightBorder;
    }
    int getLeftBorder(vector<int>& nums, int target)
    {
        int left = 0;
        int right = nums.size() - 1;
        int leftBorder = -2; // 记录一下leftBorder没有被赋值的情况
        while (left <= right) {
            int middle = left + ((right - left) / 2);
            if (nums[middle] == target) { // 寻找左边界，nums[middle] == target的时候更新right
                right = middle - 1;
                leftBorder = right + 1;
            } else if (nums[middle] < target) {
                left = middle + 1;
            } else {
                right = middle - 1;
            }
        }
        return leftBorder;
    }
};
// @lc code=end

int main()
{
    vector<int> nums = { 1 };
    Solution s;
    vector<int> res = s.searchRange(nums, 1);
    for (int i = 0; i < res.size(); i++) {
        cout << res[i] << " ";
    }
    cout << endl;
}

File: 数组/34.在排序数组中查找元素的第一个和最后一个位置.py
from typing import List


class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        left_border = self.get_left_border(nums, target)
        right_border = self.get_right_border(nums, target)
        if left_border == -1 and right_border == -1:
            return [-1, -1]
        else:
            return [left_border, right_border]

    def get_left_border(self, nums: List[int], target: int):
        """获得target的左边界
        关键在于设定一个初始的边界下标值,如果找到了新的下标,则更新,否则返回这个设定的初始下标值(-1)

        Args:
            nums (List[int]): 子数组
            target (int): 目标值

        Returns:
            _type_: _description_
        """
        left = 0
        right = len(nums) - 1
        left_border = -1
        while left <= right:
            mid = left + (right - left) // 2
            if nums[mid] < target:
                left = mid + 1
            elif nums[mid] > target:
                right = mid - 1
            elif nums[mid] == target:
                left_border = mid
                right = mid - 1
        return left_border

    def get_right_border(self, nums: List[int], target: int):
        """获得target的右边界

        Args:
            nums (List[int]): 子数组
            target (int): 目标值
        """
        left = 0
        right = len(nums) - 1
        right_border = -1
        while left <= right:
            mid = left + (right - left) // 2
            if nums[mid] < target:
                left = mid + 1
            if nums[mid] > target:
                right = mid - 1
            if nums[mid] == target:
                right_border = mid
                left = mid + 1
        return right_border


if __name__ == "__main__":
    nums = []
    print(Solution().searchRange(nums, 0))
